<a href="https://colab.research.google.com/github/kimkukhwa/M-M_Color_Distribution_Statistical_Analysis/blob/main/Part_2_M%26M_Color_Distribution_Datasets_Hypothesis_Testing_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Goodness of Fit Test
To determine whether or not an observed categorical variable follows a expected distribution.

"... checked out M&M’s web site. According to it, each package of Milk Chocolate M&M’s should contain 24% blue, 14% brown, 16% green, 20% orange, 13% red, and 14% yellow M&M’s." https://joshmadison.com/2007/12/02/mms-color-distribution-analysis/#graphs

## H0: A variable follows a hypothesized distribution.
## H1: A variable does not follow a hypothesized distribution.

In [12]:
import scipy.stats as stats
import pandas as pd
import numpy as np

In [16]:
sample_db=pd.read_csv('/content/drive/MyDrive/Personal/mm_analysis.csv',index_col=0)
sample_db.head()


In [45]:
sample=sample_db[sample_db['package_n']==48]
sample

package_n   Color  Qty observed  pack_totalQ  %ofpack  avgQ  % expected  \
282         48    Blue            14           55    25.45    55          24   
283         48   Brown             8           55    14.55    55          14   
284         48   Green            10           55    18.18    55          16   
285         48  Orange             9           55    16.36    55          20   
286         48     Red             8           55    14.55    55          13   
287         48  Yellow             6           55    10.91    55          14   

     Qty expected  Qty diff  
282          13.0       1.0  
283           8.0       0.0  
284           9.0       1.0  
285          11.0      -2.0  
286           7.0       1.0  
287           8.0      -2.0

In [27]:
expected=sample['% expected']
expected

174    24
175    14
176    16
177    20
178    13
179    14
Name: % expected, dtype: int64

In [30]:
observed=sample['%ofpack']
observed

174     8.77
175    15.79
176    17.54
177    26.32
178    12.28
179    19.30
Name: %ofpack, dtype: float64

# Chi-Square Goodness of Fit Test using the chisquare function from the SciPy library

In [39]:
#perform Chi-Square Goodness of Fit Test

print (sample_db[sample_db['package_n']==30].T.reset_index())
sample30=sample_db[sample_db['package_n']==30]
expected=sample30['% expected']
observed=sample30['%ofpack']

stats.chisquare(f_obs=observed,f_exp=expected)

          index   174    175    176     177    178     179
0     package_n    30     30     30      30     30      30
1         Color  Blue  Brown  Green  Orange    Red  Yellow
2  Qty observed     5      9     10      15      7      11
3   pack_totalQ    57     57     57      57     57      57
4       %ofpack  8.77  15.79  17.54   26.32  12.28    19.3
5          avgQ    55     55     55      55     55      55
6    % expected    24     14     16      20     13      14
7  Qty expected  14.0    8.0    9.0    11.0    7.0     8.0
8      Qty diff  -9.0    1.0    1.0     4.0    0.0     3.0


Power_divergenceResult(statistic=14.085218946886448, pvalue=0.01507706014782214)

### Result: when the ['package_n']==30, The Chi-Square test statistic is found to be 14.1 and the corresponding p-value is 0.015.

### Interpretation: Since the p-value (0.015) is less than 0.05, we reject the null hypothesis. This means we do have sufficient evidence to say that the true color distribution of M$M (package 30) is different from the distribution that the company claimed.

In [46]:
print (sample_db[sample_db['package_n']==48].T.reset_index())
sample30=sample_db[sample_db['package_n']==48]
expected=sample30['% expected']
observed=sample30['%ofpack']

stats.chisquare(f_obs=observed,f_exp=expected)

          index    282    283    284     285    286     287
0     package_n     48     48     48      48     48      48
1         Color   Blue  Brown  Green  Orange    Red  Yellow
2  Qty observed     14      8     10       9      8       6
3   pack_totalQ     55     55     55      55     55      55
4       %ofpack  25.45  14.55  18.18   16.36  14.55   10.91
5          avgQ     55     55     55      55     55      55
6    % expected     24     14     16      20     13      14
7  Qty expected   13.0    8.0    9.0    11.0    7.0     8.0
8      Qty diff    1.0    0.0    1.0    -2.0    1.0    -2.0


Power_divergenceResult(statistic=1.9355311446886447, pvalue=0.8579917092086898)

### Result: when the ['package_n']==48, The Chi-Square test statistic is found to be 1.93 and the corresponding p-value is 0.857.

### Interpretation: Since the p-value (0.015) is NOT less than 0.05, we failed to reject the null hypothesis. This means the we do NOT have sufficient evidence to say true M&M color distribution of M$M (package 48) is different than the company claimed.